kedro/framework/cli/starters.py

```diff
@@ -95,7 +97,44 @@ class KedroStarterSpec:
 KEDRO_PATH = Path(kedro.__file__).parent
 TEMPLATE_PATH = KEDRO_PATH / "templates" / "project"
 
-_STARTERS_REPO = "git+https://github.com/kedro-org/kedro-starters.git"
+
+def _get_latest_starters_version() -> str:
+    if "KEDRO_STARTERS_VERSION" not in os.environ:
+        GITHUB_TOKEN = os.getenv("GITHUB_TOKEN")
+        headers = {}
+        if GITHUB_TOKEN:
+            headers["Authorization"] = f"token {GITHUB_TOKEN}"
+
+        try:
+            response = requests.get(
+                "https://api.github.com/repos/kedro-org/kedro-starters/releases/latest",
+                headers=headers,
+                timeout=10,
+            )
+            response.raise_for_status()  # Raise an HTTPError for bad status codes
+            latest_release = response.json()
+        except requests.exceptions.RequestException as e:
+            logging.error(f"Error fetching kedro-starters latest release version: {e}")
+            return ""
+
+        os.environ["KEDRO_STARTERS_VERSION"] = latest_release["tag_name"]
+        return str(latest_release["tag_name"])
+    else:
+        return str(os.getenv("KEDRO_STARTERS_VERSION"))
+
+
+def _kedro_and_starters_version_identical() -> bool:
+    starters_version = _get_latest_starters_version()
+    return True if version == starters_version else False
+
+
+_STARTERS_REPO = (
+    "git+https://github.com/kedro-org/kedro-starters.git"
+    if _kedro_and_starters_version_identical()
+    else "https://github.com/kedro-org/kedro-starters.git@main"
```

Comment:
What happens if we point to the main branch (`_STARTERS_REPO="https://github.com/kedro-org/kedro-starters.git@main"`) but then below in `_make_cookiecutter_args_and_fetch_template` we set the checkout version as well? For example in this case:
```
if "PySpark" in tools and "Kedro Viz" in tools:
        # Use the spaceflights-pyspark-viz starter if both PySpark and Kedro Viz are chosen.
        cookiecutter_args["directory"] = "spaceflights-pyspark-viz"
        # Ensures we use the same tag version of kedro for kedro-starters
        cookiecutter_args["checkout"] = version
```
Will it then work as we expect I mean creating a project and passing the branch and the checkout version to the cookiecutter?
https://github.com/kedro-org/kedro/blob/15ba4dae69231b5f92921ddab879f4df43233e61/kedro/framework/cli/starters.py#L932

Reply:
This will lead us to cookiecutter shenanigans, which is where these parameters are ultimately used. You have to dig down four or five functions to see it, but cookiecutter is using the template parameter to determine the repo address and the checkout parameter to determine the branch, tag or commit ID. What it does is the following:

```py
    if clone:
        try:
            subprocess.check_output(  # nosec
                [repo_type, 'clone', repo_url],
                cwd=clone_to_dir,
                stderr=subprocess.STDOUT,
            )
            if checkout is not None:
                checkout_params = [checkout]
                # Avoid Mercurial "--config" and "--debugger" injection vulnerability
                if repo_type == "hg":
                    checkout_params.insert(0, "--")
                subprocess.check_output(  # nosec
                    [repo_type, 'checkout', *checkout_params],
                    cwd=repo_dir,
                    stderr=subprocess.STDOUT,
                )
```

So if you passed the repo_url as the address to the main branch and, for example, passed the checkout value as a different branch, cookiecutter would clone the main branch and then checkout to the other branch.

Reply:
Thank you for explaining, that's what I thought as well. Doesn't it mean that in this case instead of `main` branch (which we aim for) we again will have the released branch (as the checkout argument is the installed kedro version in case it's not passed https://github.com/kedro-org/kedro/blob/15ba4dae69231b5f92921ddab879f4df43233e61/kedro/framework/cli/starters.py#L344) which is not desired?

Reply:
That's a good point, let me test that.

Reply:
I was thinking of refactoring the logic in `_make_cookiecutter_args_and_fetch_template` because now we have two places where we define the template which is a bit confusing. But that can be done in a separate PR.

Now, we can just identify the case when repos versions are not matched (as you do for setting the `_STARTERS_REPO`). In this case, we only need to use the template path provided but skip setting up the checkout parameter in `_make_cookiecutter_args_and_fetch_template`.

Reply:
Or we can just do something like 
```
checkout = checkout or version if _kedro_and_starters_version_identical() else None
```

https://github.com/kedro-org/kedro/blob/15ba4dae69231b5f92921ddab879f4df43233e61/kedro/framework/cli/starters.py#L399